### ISSUE TITLE
RPG Shooter: Starwish Prologue softlock

### ISSUE THREAD

**Issue by waspennator** [ruffle-rs/ruffle#14932]
### Describe the bug

When playing through the prologue shmup part of the game at the start, you'll quickly run into a softlock when you are forced to lose to the first boss ship of the game, some assets stopping once you lose all your health, but not everything.

Just start a new game and skip through everything till you get to the shmup part of the game, shouldn't take longer than a minute to reach the boss.

### Expected behavior

The game should transition into the next cutscene when you die to the fireballs boss ship.

### Content Location

https://www.newgrounds.com/portal/view/572374?emulate=flash

### Affected platform

Browser's extension

### Operating system

Android 13

### Browser

Chrome 120

**Comment by heavens-thesis** [comment]
It's not a Ruffle bug, it's a Starwish bug--at that point the game is told to update your stats through the Kongregate API (no matter what site you're playing on) and there's nothing in the code to handle a failure to do so. This will happen on any emulator or flash projector without wider internet connectivity and will someday happen no matter what once Kongregate changes its API or shuts down.

**Comment by Sworddragon** [comment]
> at that point the game is told to update your stats through the Kongregate API (no matter what site you're playing on) and there's nothing in the code to handle a failure to do so.

But the game worked in the past (when Adobe Flash was still alive) on other sites. Unless the cross-site usage of the Kongregate API did not trigger a failure here.


> This will happen on any emulator or flash projector without wider internet connectivity and will someday happen no matter what once Kongregate changes its API or shuts down.

It already happens for 1 or 2 years even when being played on Kongregate. If this is really not an emulation bug I assume Kongregate has already made changes to their API that breaks backwards compatibility for some games. Though, while this should be unusual I wouldn't be surprised if Kongregate did this in the past few years. But so far I don't recall having seen any non-flash game breaking as well but I might just have been lucky.


I assume the failure throws an exception that causes a halt of the current code context and thus softlocking the game (and thus an old browser/flash setup would just experience the same)?


I'm throwing this thoughts in with a bit of a hope that there might have been an oversight as otherwise it means the game is currently already unplayable for anybody no matter the method it is being played.

**Comment by heavens-thesis** [comment]
> But the game worked in the past (when Adobe Flash was still alive) on other sites. Unless the cross-site usage of the Kongregate API did not trigger a failure here.

Correct, the game itself does so on frame 1 regardless of which gaming site you get the game from. You could grab it from onemorelevel.com, or Newgrounds, or even the author's own Deviantart page, and it will always have the api path setup and such in the code:

```
this.apiPath = this.paramObj.kongregate_api_path || "http://www.kongregate.com/flash/API_AS3_Local.swf";

Security.allowDomain(this.apiPath);

this.request = new URLRequest(this.apiPath);
```
> I assume the failure throws an exception that causes a halt of the current code context and thus softlocking the game (and thus an old browser/flash setup would just experience the same)?

I don't know what that means, so I can't confirm or deny it, though I will say that any downloaded .swf of the game containing the api calls will also fail to run in the official Adobe standalone Flash projector.
> I'm throwing this thoughts in with a bit of a hope that there might have been an oversight...

Maybe there has been an oversight, my understanding isn't perfect.
> ... as otherwise it means the game is currently already unplayable for anybody no matter the method it is being played.

Untrue, the relevant parts of the code can be taken out and it will run just fine in a standalone projector. Similarly, the site-exclusive-content gates can be removed to access Neferiti's content or the (internally named) "Harem" ending. Of course, the player would still be on the hook for meeting the requirements themselves and not triggering the anticheat. [Here](https://mega.nz/file/fPZTzboT#BN3WRrviSxwhCNxp2TJdp7YVXsBL4vbMcFOXC1plbHA) is a proof of concept.

**Comment by Sworddragon** [comment]
On testing this again I was able to beat the boss and the game just continued without the softlock. But on repeating this by intentionally losing the game did just softlock again.

So a workaround for the time being is to just defeat the boss but this is usually not an easy task as the game intends you to lose here. And as a site-note there are a few bugs afterwards (mostly small glitches like your ship visually duplicating every stage clear or numbers in the hangar are not updating until entering the hangar again, but there is also a softlock once you enter the weapon shop).

**Comment by heavens-thesis** [comment]
Once again, not a ruffle bug. The game tries to comtact Kongregate to update your weapon level score when you enter the weapon shop. The same thing will happen if you avoid weapons somehow and enter the mods shop. The same will happen if you die to anything else.

Once again! The exact same bug happens in any flash projector that doesn't let out interactions to the internet. For example, Adobe's final standalone official projector. There's no point trying to follow up on this through Ruffle because the needed functionality is not in Ruffle's scope.
